custom/logistics/tasks.py

```
@@ -101,3 +113,98 @@ def add_products_to_loc(api):
+        while has_next:
+            meta, stocktransactions = endpoint.get_stocktransactions(next_url_params=next_url,
+                                                                     limit=limit,
+                                                                     offset=offset,
+                                                                     filters=(dict(supply_point=supply_point,
+                                                                                   date__gte=date,
+                                                                                   order_by='date')))
+
+            # set the checkpoint right before the data we are about to process
+            save_stock_data_checkpoint(checkpoint,
+                                       'stock_transaction',
+                                       meta.get('limit') or limit,
+                                       meta.get('offset') or offset,
+                                       date, facility, True)
+            transactions_to_add = []
+            with transaction.commit_on_success():
+                for stocktransaction in stocktransactions:
+                    params = dict(
+                        form_id=xform._id,
+                        date=force_to_datetime(stocktransaction.date),
+                        type='balance',
+                        domain=domain,
+                    )
+                    try:
+                        report, _ = StockReport.objects.get_or_create(**params)
+                    except StockReport.MultipleObjectsReturned:
+                        # legacy
+                        report = StockReport.objects.filter(**params)[0]
+                    try:
+                        sql_product = SQLProduct.objects.get(code=stocktransaction.product, domain=domain)
+                    except SQLProduct.DoesNotExist:
+                        # todo: kkrampa what's the deal with this logic? this should never be true
```

if this should never be true maybe we should throw an error instead of continuing? or at least some sort of notification so we know something has been corrupted